# Funciones recursivas 

## Introducción

__¿Cómo entrega los regalos el Viejito Pascuero?__

Suponga que, el viejito pascuero:

1. tiene una lista de las casas por las que pasa.
2. pasa por una casa (entra por la chimenea) y deja los regalos.

Una representación del proceso por medio de un algoritmo, se basaría en una estructura repetitiva, por ejemplo:

In [2]:
def entrega_regalos(casas):
    for casa in casas:
        print('Entregando regalo a {}'.format(casa))

In [3]:
casas = ['Manuela', 'Juana', 'Alexis', 'José']
entrega_regalos(casas)

Entregando regalo a Manuela
Entregando regalo a Juana
Entregando regalo a Alexis
Entregando regalo a José


Suponga ahora, que el Viejito Pascuero tiene a su disposición un conjunto de ayudantes que le ayudarán a entregar los regalos y, por lo tanto, puede divir el trabajo entre ellos. A su vez, cada uno de ellos puede dividir su trabajo entre otros ayudantes.

![Diagrama entrega regalos](./img/recursividad-ejemplo-regalos.svg)

La estructura anterior, es una estructura recursiva:
- Si el caso es simple, se puede resolver.
- De lo contrario, el problema se divide en subproblemas y se aplica la misma estrategia.

A continuación se presenta un algoritmo para resolver el problema anterior, aplicando recursividad:

In [4]:
def entrega_regalos_recursiva(casas):
    # el ayudante hace su trabajo
    if len(casas) == 1:
        print('Entregando relago a {}'.format(casas[0]))
    
    else:
        mitad = len(casas) // 2
        primera_mitad = casas[:mitad]
        segunda_mitad = casas[mitad:]
        
        # el ayudante divide su trabajo
        entrega_regalos_recursiva(primera_mitad)
        entrega_regalos_recursiva(segunda_mitad)

In [5]:
entrega_regalos_recursiva(casas)

Entregando relago a Manuela
Entregando relago a Juana
Entregando relago a Alexis
Entregando relago a José


## Definición formal

Una función recursiva es __una función definida en terminos de que se invoque a si misma__. Las funciones recursivas comparten la siguiente estructura en comun:

- Debe tener un __condición base__.
- Debe __orientar su lógica para alcanzar la condición base__.
- Debe __invocarse a si misma__.

A medida que se realizan llamadas a funciones, __el estado anterior de las funciones__ se almacena en una pila de _llamadas a funciones_:
- Esto permite que sea posible que una función se llame a sí misma, y
- Que las variables dentro de la función tomen distintos valores.

## Ejemplo: Función factorial



El __factorial__ de un número entero positivo $n$, se define como el producto de todos los números menores o iguales:

$$ n! = \prod\limits_{k=1}^{n} k = 1 \times 2 \times 3 \times \dots \times (n - 1) \times n $$

Descomponiendo el problema original en simples instancias del mismo problema:

\begin{align}
n! &= n \times (n-1)!\\
n! &= n \times (n-1) \times (n-2)!\\
n! &= n \times (n-1) \times (n-2) \times (n-3)!\\
\vdots& \vdots \\
n! &= n \times (n-1) \times (n-2) \times (n-3) \dots \times 3!\\
n! &= n \times (n-1) \times (n-2) \times (n-3) \dots \times 3 \times 2!\\
n! &= n \times (n-1) \times (n-2) \times (n-3) \dots \times 3 \times 2 \times 1!\\
\end{align}

A medida que __el problema se descompone__ en problemas sucesivamente menos complejos, estos deben llegar a ser tan simples que pueda __resolverse sin necesidad de subdividirlos más__ (caso base). El caso base para el ejemplo:

\begin{align}
1! = 1
\end{align}

En consecuencia, la función recursiva para calcular $n!$:

In [7]:
def factorial_recursiva(n):
    if n <= 1:
        return 1
    else:
        return n * factorial_recursiva(n-1)

In [8]:
factorial_recursiva(5)

120

La invocación de la función `factorial_recursiva(5)`, genera la siguiente ejecución,

| Invocación | Contexto | | |
| :-- | :-- | -- | -- |
| `factorial_recursiva(5)` | `5 * factorial_recursiva(4)` |	↓ |	`5 * 24 = 120`|
| `factorial_recursiva(4)` | `4 * factorial_recursiva(3)` |	  | `4 * 6 = 24`|
| `factorial_recursiva(3)` | `3 * factorial_recursiva(2)` |	  | `3 * 2 = 6`|
| `factorial_recursiva(2)` | `2 * factorial_recursiva(1)` |	  | `2 * 1 = 2`|
| `factorial_recursiva(1)` | `1`                          | ↗ |  |

- En cada llamada recursiva se añade una pila que contiene su contexto de ejecución hasta alcanzar la condición base.
- Luego se comienza a calcular a medida que se retorna el resultado.

## Ejemplo: Función potenciación

La potenciación de una base ($a$) y un exponente perteneciente al conjunto de los números naturales ($n$), se define como la multiplicación de la base $n$ veces:

\begin{align}
a^{n} = a_{1} \times a_{2} \times \dots \times a_{n}
\end{align}

Una función iterativa de la operación potenciación, se implementa de la siguiente forma:

In [11]:
def potencia(base, exponente):
    if exponente == 0:
        return 1
    else:
        ans = 1
        while exponente > 0:
            print(exponente)
            ans *= base
            exponente -= 1
    return ans

In [12]:
potencia(2,10)

10
9
8
7
6
5
4
3
2
1


1024

Considerando las siguientes propiedades,

- Si $n$ es par:

\begin{equation}
a^n = a^{\frac{n}{2}} \times a^{\frac{n}{2}}
\end{equation}

- Si $n$ es impar:

\begin{equation}
a^n = a^{\frac{(n-1)}{2}} \times a^{\frac{(n-1)}{2}}\times a
\end{equation}

En consecuencia, una función recursiva para la operación potenciación, se implementaría de la siguiente forma:

In [7]:
def potencia_recursiva(base, exponente):
    # restricción: para exponente >= 0
    # caso base: si exponente == 0 => 1
    if exponente <= 0:
        return 1
    else:
        # para exponente par
        if exponente % 2 == 0:
            ans = potencia_recursiva(base, exponente/2)
            return ans * ans
        # para exponente impar
        else:
            ans = potencia_recursiva(base, (exponente-1)/2)
            return ans * ans * base

In [14]:
potencia_recursiva(2, 10)

1024

La invocación de la función `potencia_recursiva(2,10)`, genera la siguiente ejecución,

| Invocación | ans | Contexto |     |     |
| :--        | :-- | :--      |:--  |:--  |
| `potencia_recursiva(2,10)` | `potencia_recursiva(2,10/2)`    | `ans * ans`     | ↓ | `32 * 32 = 1024` |
| `potencia_recursiva(2,5)`  | `potencia_recursiva(2,(5-1)/2)` | `ans * ans * 2` |   | `4 * 4 * 2 = 32` |
| `potencia_recursiva(2,2)`  | `potencia_recursiva(2,2/2)`     | `ans * ans`     |   | `2 * 2 = 4`      |
| `potencia_recursiva(2,1)`  | `potencia_recursiva(2,(1-1)/2)` | `ans * ans * 2` |   | `1 * 1 * 2 = 2`  |
| `potencia_recursiva(2,0)`  | `1`                             |                 | ↗ |                  |

En ejemplo anterior,
- La versión iterativa realiza ```n = exp = 10``` iteraciones.
- La versión recursiva realiza ```5``` invocaciones.

En general, es importante notar que, 
- La __versión recursiva es más elegante y legible__ que la iterativa,
- En la versión recursiva, __existe una estructura condicional__ (condición base).
- Para toda función recursiva __es posible encontrar otra que realice el mismo cálculo de forma iterativa__.
- En la versión iterativa, existe un __ciclo repetitivo__.